hermes/probe/check_nil.go

```diff
@@ -0,0 +1,159 @@
+// It then checks that the file has been deleted by trying to get the object.
+// Arguments:
+//	- ctx: pass the context so this probe can be cancelled if needed.
+//	- target: pass the target run information.
+//	- client: pass an initialised storage client for this target system.
+//	- logger: pass the logger associated with the probe calling this function.
+// Returns:
+//	- err:
+//		- [...].hermes_file_missing: the file to be deleted does not exist in Hermes' StateJournal.
+//		- [...].file_missing: the file to be deleted could not be found in the target bucket.
+//		- [...].bucket_missing: the target bucket on this target system was not found.
+//		- [...].probe_failed: there was an error during one of the API calls and the probe failed.
+//		- [...].deleted_file_found: the file was deleted but it is still found in the target bucket.
+//		- [...].list_bucket_failed: the listBucket operation failed when checking if the target file was deleted.
+// TODO(evanSpendlove): Add exit status to return values.
+func CheckNilFile(ctx context.Context, target *Target, client *stiface.Client, logger *logger.Logger) (m.ExitStatus, error) {
+	bucket := target.Target.GetBucketName()
+
+	// TODO(evanSpendlove): Find the Nil file on the storage system, alert if not present.
+	query := &storage.Query{Prefix: "Hermes_NIL"}
+	start := time.Now()
+	objIter := (*client).Bucket(bucket).Objects(ctx, query) // Can add API call timeouts here.
+	end := time.Now()
+	var nilFilename string
+	for {
+		obj, err := objIter.Next()
+		if err != nil {
+			target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_LIST_FILES]][m.ExitStatuses[m.FILE_MISSING]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+			return fmt.Errorf("checkNilFile.list_files_failed: unable to find nil file: %w", err)
+		}
+		nilFilename = obj.Name
+	}
+	target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_LIST_FILES]][m.ExitStatuses[m.SUCCESS]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+	logger.Infof("checkNilFile: nil file located on target storage system.")
+
+	// TODO(evanSpendlove): Get the contents of the NIL file.
+	nilFile := (*client).Bucket(bucket).Object(nilFilename)
+
+	start = time.Now()
+	reader, err := nilFile.NewReader(ctx)
+	end = time.Now()
+	if err != nil {
+		target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_GET_FILE]][m.ExitStatuses[m.FILE_READ_FAILURE]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+		return fmt.Errorf("checkNilFile.read_file_failed: unable to read nil file: %w", err)
+	}
+	target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_GET_FILE]][m.ExitStatuses[m.SUCCESS]].Metric("latency").AddFloat64(end.Sub(start).Seconds())
+	logger.Infof("checkNilFile: nil file gotten from target storage system.")
+
+	// TODO(evanSpendlove): Amend metric recording here to use api_READ or probe op read and verify metrics.
+	start = time.Now()
+	nilContents, err := ioutil.ReadAll(reader)
```

Comment:
It is necessary to retain the contents in memory, just not for the `sha1` function. They are used later on in the `proto.Unmarshal()` call, which does not accept a Reader as it does not work with streams.

Also, it would be very unlikely that this file would be larger than the available memory as it only ever contain the serialized Journal proto. This proto has a maximum implemented capacity of 50 map[int32]string entries and 1 Intent, so it will be quite small (<2KB).